src/physics/barneshut.rs

```diff
@@ -0,0 +1,149 @@
+use geometry::types::Point;
+use geometry::types::Rect;
+use geometry::types::Quadrant::{NW, NE, SW, SE};
+use super::types::Body;
+
+// RealBody //////////////////////////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub struct RealBody {
+    id: usize,
+    mass: f32,
+    position: Point,
+}
+
+impl RealBody {
+    // TODO: validate
+    fn new(id: usize, mass: f32, x: f32, y: f32) -> RealBody {
+        RealBody {
+            id,
+            mass,
+            position: Point { x, y },
+        }
+    }
+}
+
+// BHNode ////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+pub struct BHNode {
+    space: Rect,
+    body: Option<RealBody>,
+    nw: Option<Box<BHNode>>,
+    ne: Option<Box<BHNode>>,
+    sw: Option<Box<BHNode>>,
+    se: Option<Box<BHNode>>,
+}
+
+impl BHNode {
+    fn new(space: Rect) -> Self {
+        BHNode {
+            space,
+            body: None,
+            nw: None,
+            ne: None,
+            sw: None,
+            se: None,
+        }
+    }
+
+//    fn body(&self) -> RealBody {
+//        // we need to find the center of mass for this node.
+//        RealBody {
+//            id: 0,
+//            mass: 0.0,
+//            position: Point::origin(),
+//        }
+//    }
+
+    // TODO: needs testing
```

Comment:
i saw lots of `// TODO:` in the code, could you please explain for me which function is this little code is used for ?

Reply:
It’s just a reminder that I need to do somethinf